lib/services/devices/deviceService.js

```diff
@@ -116,13 +187,119 @@ function getInitialValueForType(type) {
+ * Concats or merges two JSON objects.
+ *
+ * @param  {Object} json1           JSON object where objects will be merged.
+ * @param  {Object} json2           JSON object to be merged.
+ */
+function jsonConcat(json1, json2) {
+    for (var key in json2) {
+        if (json2.hasOwnProperty(key)) {
+            json1[key] = json2[key];
+        }
+    }
+}
+
+/**
+ * Formats device's attributes in NGSIv2 format.
+ *
+ * @param  {Object} originalVector  Original vector which contains all the device information and attributes.
+ * @param  {Object} staticAtts      Flag that defined if the device'attributes are static.
+ * @return {Object}                 List of device's attributes formatted in NGSIv2.
+ */
+function formatAttributesNgsi2(originalVector, staticAtts) {
+    var attributeList = {};
+
+    if (originalVector && originalVector.length) {
+        for (var i = 0; i < originalVector.length; i++) {
+            attributeList[originalVector[i].name] = {
+                type: originalVector[i].type,
+                value: getInitialValueForType(originalVector[i].type)
+            };
+
+            if (staticAtts) {
+                attributeList[originalVector[i].name].value = originalVector[i].value;
+            } else {
+                attributeList[originalVector[i].name].value = getInitialValueForType(originalVector[i].type);
+            }
+        }
+    }
+
+    return attributeList;
+}
+
+/**
+ * Formats device's commands in NGSIv2 format.
+ *
+ * @param  {Object} originalVector  Original vector which contains all the device information and attributes.
+ * @return {Object}                 List of device's commands formatted in NGSIv2.
+ */
+function formatCommandsNgsi2(originalVector) {
+    var attributeList = {};
+
+    if (originalVector && originalVector.length) {
+        for (var i = 0; i < originalVector.length; i++) {
+            attributeList[originalVector[i].name + constants.COMMAND_STATUS_SUFIX] = {
+                type: constants.COMMAND_STATUS,
+                value: 'UNKNOWN'
+            };
+            attributeList[originalVector[i].name + constants.COMMAND_RESULT_SUFIX] = {
+                type: constants.COMMAND_RESULT,
+                value: ' '
+            };
+        }
+    }
+
+    return attributeList;
+}
+
+/**
+ * Creates the initial entity representing the device in the Context Broker using NGSIv2.
+ * This is important mainly to allow the rest of the updateContext operations to be performed.
  *
  * @param {Object} deviceData       Object containing all the deviceData needed to send the registration.
  * @param {Object} newDevice        Device object that will be stored in the database.
  */
-function createInitialEntity(deviceData, newDevice, callback) {
+function createInitialEntityNgsi2(deviceData, newDevice, callback) {
+    var options = {
+        url: config.getConfig().contextBroker.url + '/v2/entities',
+        method: 'POST',
```

Comment:
I have realized that this could be problematic...

It is a common use case that the entity used by the IOTA agent is "shared" among context based applications. Typically, using a combination of "low level" attributes (mapping to phisical measure from the IOTA) and "high level" attributes (used mainly at the application layer). In some ocassions it may happen that even in the "low level" set, different attributes correspond to different devices (for instance, "tempature" is measured by a device while "pressure" is measured by another).

Thus, it may  happen that the entity the IOTA agent is attemping to create has been already creatd by another context producer applicationt. In that case, the IOTA will get a 422 at this point and the process will fail.

NGSIv1 solves this situation using `POST /v1/updateContext` with "APPEND" actionType. There is a "parallel" in NGSIv2 (POST /v2/op/update with "append" actionType. However, that operation is more complex from a syntactical point of view and though for another use case in NGSIv2, i.e. bulk updates use case in particular (this is analyzed in https://github.com/telefonicaid/fiware-orion/issues/3215).

Thus for "single entity upsert" a new operation `POST /v2/entities?options=upsert` has been defined in the NGSIv2 API. The PR to include it in Orion is on the way (https://github.com/telefonicaid/fiware-orion/pull/3217) and I think it should be the operation to use at entity creation step in IOTAs.

Reply:
Ok @fgalan. I can do the implementation right now in the iotagent-node-lib and modify the corresponding CB mocks. BTW, when do you think that https://github.com/telefonicaid/fiware-orion/pull/3217 will be merged?

Reply:
It should be merged soon... one of the two required LGTMs on it has been already provided. Only pending of @jmcanterafonseca 's one.

Reply:
Fixed with dff9f3354975812594445530e97febae597ab558

Reply:
@fgalan Is this now correct?

Reply:
Looks ok to me but better to wait until https://github.com/telefonicaid/fiware-orion/pull/3217 gets merged (@jmcanterafonseca has tool me that he will have look no later than tomorrow)